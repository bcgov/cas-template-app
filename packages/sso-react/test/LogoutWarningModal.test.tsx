import React from "react";
import LogoutWarningModal from "../src/LogoutWarningModal";
import { mount, shallow } from "enzyme";
import { act } from "react-dom/test-utils";

afterEach(() => {
  jest.useRealTimers();
});

describe("The Logout Warning Modal", () => {
  it("Should match the snapshot", () => {
    jest.useFakeTimers();
    const dateMock = jest.spyOn(Date, "now").mockImplementation(() => 1000); // One second after January 1st, 1970

    const componentUnderTest = shallow(
      <LogoutWarningModal
        onExtendSession={() => {}}
        expiresOn={12000}
        logoutPath="/logout"
      />
    );
    expect(componentUnderTest).toMatchSnapshot();

    dateMock.mockRestore();
  });

  it("Should use the optional render method if provided", () => {
    jest.useFakeTimers();
    const dateMock = jest.spyOn(Date, "now").mockImplementation(() => 1000); // One second after January 1st, 1970

    const componentUnderTest = mount(
      <LogoutWarningModal
        onExtendSession={() => {}}
        expiresOn={12000}
        logoutPath="/logout"
        renderModal={({ remainingSeconds }) => (
          <div>There are {remainingSeconds} seconds left.</div>
        )}
      />
    );
    expect(componentUnderTest).toMatchSnapshot();

    dateMock.mockRestore();
  });

  it("should call the extendSession function", () => {
    const expireSpy = jest.fn();
    const componentUnderTest = shallow(
      <LogoutWarningModal
        onExtendSession={expireSpy}
        expiresOn={12345}
        logoutPath="/logout"
      />
    );
    componentUnderTest
      .find("Button#logout-warning-modal-remain-active-button")
      .simulate("click");
    expect(expireSpy).toHaveBeenCalled();
  });

  it("should countdown seconds", async () => {
    jest.useFakeTimers();
    let dateMock = jest.spyOn(Date, "now").mockImplementation(() => 1000); // 1 second after Jan. 1st, 1970

    let componentUnderTest: any = {};

    await act(async () => {
      componentUnderTest = mount(
        <LogoutWarningModal
          onExtendSession={() => {}}
          expiresOn={17000} // 17 seconds after Jan. 1st, 1970
          logoutPath="/logout"
        />
      );
    });
    await componentUnderTest.update();

    expect(componentUnderTest).toMatchSnapshot();
    expect(
      componentUnderTest
        .text()
        .trim()
        .includes("You will be logged out in 16 seconds")
    ).toBeTrue();

    dateMock.mockRestore();
    dateMock = jest.spyOn(Date, "now").mockImplementation(() => 5000); // 5 second after Jan. 1st, 1970

    await act(async () => {
      jest.runOnlyPendingTimers();
    });

    // When the timer expires the component re-syncs from the system time
    expect(
      componentUnderTest
        .text()
        .trim()
        .includes("You will be logged out in 12 seconds")
    ).toBeTrue();

    dateMock.mockRestore();
  });
});
