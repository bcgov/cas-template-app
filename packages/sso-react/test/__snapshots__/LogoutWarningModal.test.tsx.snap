// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Logout Warning Modal Should match the snapshot 1`] = `
<BaseComponent
  id="logout-warning-modal"
  size="lg"
  style={
    Object {
      "zIndex": 100,
    }
  }
>
  <Component
    className="h4"
  >
    Inactivity Logout Warning
  </Component>
  <Component
    style={
      Object {
        "padding": "2em",
      }
    }
  >
    <BaseComponent
      cols={12}
    >
      <Component>
        Your session is about to expire due to inactivity.
      </Component>
      <Component>
        You will be logged out in 
        11
         seconds.
      </Component>
      <Component
        justify="end"
        style={
          Object {
            "marginTop": "1em",
          }
        }
      >
        <form
          action="/logout"
          method="post"
          style={
            Object {
              "marginBottom": 0,
            }
          }
        >
          <Button
            id="logout-warning-modal-logout-button"
            type="submit"
            variant="secondary"
          >
            Logout
          </Button>
        </form>
        <Button
          id="logout-warning-modal-remain-active-button"
          onClick={[Function]}
          style={
            Object {
              "marginLeft": "1em",
            }
          }
          variant="primary"
        >
          Remain active
        </Button>
      </Component>
    </BaseComponent>
  </Component>
</BaseComponent>
`;

exports[`The Logout Warning Modal Should use the optional render method if provided 1`] = `
<div>
  There are 
  11
   seconds left.
</div>
`;

exports[`The Logout Warning Modal should countdown seconds 1`] = `
<div
  className="sc-iCfMLu sc-furwcr kQXSYc gDlUxH pg-modal-container"
  id="logout-warning-modal"
  size="lg"
  style={
    Object {
      "zIndex": 100,
    }
  }
>
  <div
    className="sc-pVTFL fdkSXM pg-modal-main"
  >
    <header
      className="sc-jrQzAO eeqVZV pg-modal-header h4"
    >
      Inactivity Logout Warning
    </header>
    <div
      className="sc-kDTinF cIwtan pg-modal-content"
      style={
        Object {
          "padding": "2em",
        }
      }
    >
      <div
        className="sc-fFeiMQ bgyiGU pg-grid-container"
        cols={12}
      >
        <div
          className="sc-bkkeKt sc-ieecCq hOEAKj fEyLc pg-grid-row"
        >
          Your session is about to expire due to inactivity.
        </div>
        <div
          className="sc-bkkeKt sc-ieecCq hOEAKj fEyLc pg-grid-row"
        >
          You will be logged out in 
          16
           seconds.
        </div>
        <div
          className="sc-bkkeKt sc-ieecCq eHeiPc fEyLc pg-grid-row"
          style={
            Object {
              "marginTop": "1em",
            }
          }
        >
          <form
            action="/logout"
            method="post"
            style={
              Object {
                "marginBottom": 0,
              }
            }
          >
            <button
              className="sc-kfPuZi jgizTO pg-button"
              id="logout-warning-modal-logout-button"
              type="submit"
            >
              Logout
            </button>
          </form>
          <button
            className="sc-kfPuZi epUors pg-button"
            id="logout-warning-modal-remain-active-button"
            onClick={[Function]}
            style={
              Object {
                "marginLeft": "1em",
              }
            }
          >
            Remain active
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
